Program v3.1.0 (May Audit)

acamill opened the pull request:
### Fixed

- `Depository.getDeltaNeutralPositionNotionalSizeUI(mango: Mango)` now retrieves Mango account, PERP market & price concurrently.
- `Depository.getUnrealizedPnl(mango: Mango, options: ConfirmOptions)` now retrieves depository account & delta-neutral position size concurrently.
- `Depository.getFundingRate(mango: Mango)` now retrieves Mango Perp market & cache + bids & asks concurrently.
- `Depository.settleMangoDepositoryMangoAccountPnl(payer: Payer, mango: Mango)` now retrieves Mango account, cache, PERP market & quote root bank concurrently.

### Added

- Add `Client.createQuoteMintWithMangoDepositoryInstruction` method.
- Add `Client.createQuoteRedeemWithMangoDepositoryInstruction` method.
- Add `Client.createSetMangoDepositoryQuoteMintAndRedeemFeeInstruction` method.
- Add `Client.createSetMangoDepositoryQuoteMintAndRedeemSoftCapInstruction` method.
- Add `Client.createDisableDepositoryMintingInstruction` method.
- Add `Depository.getOffsetUnrealizedPnl(mango: Mango, options: ConfirmOptions)` method.
- Add `Depository.getCollateralPerpPriceNativeQuotePerNativeBase(mango: Mango): Promise<I80F48>` method.
- Add `Depository.getLimitPrice(slippage: I80F48, perpOrderTakerSide: 'short' | 'long', mango: Mango): Promise<I80F48> method.`
- Export new `findMultipleATAAddSync(wallet: PublicKey, mintAddresses: PublicKey[]): [PublicKey, number][]` utility function.

### Changed

- Update IDL to match UXD Program version 3.1.0 (audited).
- `Client.createRebalanceMangoDepositoryLiteInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrieve, compute & enforce the slippage limit price.
- `Client.createMintWithMangoDepositoryInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrive, compute & enforce the slippage limit price.
- `Client.createRedeemFromMangoDepositoryInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrieve, compute & enforce the slippage limit price.
- `Depository.getMinTradingSizeCollateralUI(mango: Mango)` now returns a `number` synchronously.
- `getBalance` utility function now throws an `Error` instead of a `string` when unable to retrieve token account balance.
- Upgrade `@project-serum/anchor` to `0.24.2` (from `0.22.0`).
- Upgrade `@solana/web3.js` to `1.42.0` (from `1.35.1`).
- Add `mangoDepositoriesQuoteRedeemableSoftCap!: BN; // u64` property to the Controller account.
- Add the following properties to the Depository account:
```
  netQuoteMinted!: BN; // i128  // The signed amount of quote-minted redeemable-tokens, negative means redeemed
  quoteMintAndRedeemFee!: number; // u8 // The fee applied to quote-mint/redeem
  totalQuoteMintAndRedeemFees!: BN; // u128 // The total amount of fee collected
  mintingDisabled!: boolean; // If the depository has disabled both collateral and quote minting
```
- All token account required by program instructions now need to be created by the caller. The program does not initialise accounts on behalf of the user anymore

### Removed

- Remove `findAddrSync` utility function export, use `PublicKey.findProgramAddressSync` instead.

acamill committed:
WIP: feat(01): Add 01 (ZeroOne) DEX / Protocol support

- Add `01` folder with ZeroOne protocol related code similar to what is done with MangoMarkets
- BREAKING: Some transaction creation instructions now return a promise instead - I think this is bad because it's now blocking? What is the best way, inject the needed accounts?
- BREAKING: Upgrade anchor to `0.24.2` (from `0.22.0`)
- Add methods to retrieve Pnl for borrow/deposit/funding from Mango Stats in MangoDepository
- Remove Rent/Associated program, all accounts must be init before passed to instructions now, no more init if needed

acamill commented:
@0xCryptoSheik  How to handle these issues with zero_one dep, can we ignore?
<img width="778" alt="Screenshot 2022-05-06 at 10 58 53" src="https://user-images.githubusercontent.com/40299747/167059637-e9fdd2ff-6e70-42e3-8b16-211e228afa34.png">

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+        zoState = ZO_MAINNET_STATE_KEY;
+        break;
+      default: {
+        throw 'error cluster';
```
`throw new Error('Invalid cluster')`

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+  ): Promise<ZoDepositoryAccount> {
+    const uxdProgram = new Program(IDL, this.uxdProgramId);
+
+    const ai = await connection.getAccountInfo(this.pda, options?.commitment);
```
`ai` ? Use a more descriptive variable name?

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+
+  getPerpPriceNum(zo: Zo): Num {
+    const pmi = zo.state.getMarketIndexBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
nit: \`${this.collateralMintSymbol}-PERP\`

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+    controlKey: PublicKey
+  ): Promise<PublicKey> {
+    const dexMarket = await zo.state.getMarketBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
same as above, use template string

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+
+  async minTradingSizeBase(zo: Zo): Promise<number> {
+    const dexMarket = await zo.state.getMarketBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
same as above, use template string

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+
+  public async settleMarginAccountFunds(zo: Zo): Promise<string> {
+    let zoAccount = await zo.loadMarginAccount(this.zoAccountPda);
+    return zoAccount.settleFunds(this.collateralMintSymbol + '-PERP');
```
same as above, use template string

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+    return zoAccount.settleFunds(this.collateralMintSymbol + '-PERP');
+  }
+
+  // async minTradingSizeQuote(zo: Zo): Promise<number> {
```
remove before merge if not included in this version

0xCryptoSheik replied:
@acamill keep & finish or remove?

0xCryptoSheik commented on src/01/index.ts:
```diff
@@ -0,0 +1,81 @@
+const invalidClusterError = 'Invalid cluster';
+
+// Helper to init and configure a Zo object.
+export async function createAndInitializeZo(
```
remove `async`

0xCryptoSheik commented on src/01/index.ts:
```diff
@@ -0,0 +1,81 @@
+} from '@zero_one/client';
+import * as anchorZo from '@zero_one/client/node_modules/@project-serum/anchor';
+
+const invalidClusterError = 'Invalid cluster';
```
inline if only used once, wrap in a `new Error(...)`

0xCryptoSheik commented on src/01/index.ts:
```diff
@@ -0,0 +1,81 @@
+  Margin as zoMargin,
+  Zo as ZoIdl,
+} from '@zero_one/client';
+import * as anchorZo from '@zero_one/client/node_modules/@project-serum/anchor';
```
Uh oh!
Need to try align these.

0xCryptoSheik commented on src/01/index.ts:
```diff
@@ -0,0 +1,81 @@
+    state.startCacheRefreshCycle();
+  }
+
+  public async loadMarginAccount(owner: PublicKey): Promise<zoMargin> {
```
remove `async`

0xCryptoSheik commented on src/01/index.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  public async loadMarginAccount(owner: PublicKey): Promise<zoMargin> {
+    return await zoMargin.load(
```
remove `await`

0xCryptoSheik commented on src/client.ts:
```diff
@@ -128,12 +126,15 @@ export class UXDClient {
-    )[0];
-    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
+  ): Promise<TransactionInstruction> {
+    const userCollateralATA = await utils.token.associatedAddress({
```
wrap in a `Promise.all` since both can be done concurrently

0xCryptoSheik commented on src/client.ts:
```diff
@@ -160,13 +161,19 @@ export class UXDClient {
+    const maxRebalancingAmountNative = uiToNative(maxRebalancingAmount, depository.quoteMintDecimals);
+    const perpSide = polarity == PnLPolarity.Positive ? 'short' : 'long'; //'sell' : 'buy';
+    const limit_price = (
+      await depository.getLimitPrice(
```
Ideally this one too in the `Promise.all` of above but let's keep it simple for now

0xCryptoSheik replied:
also, camelcase

0xCryptoSheik commented on src/client.ts:
```diff
@@ -160,13 +161,19 @@ export class UXDClient {
-      maxRebalancingAmount * 10 ** depository.quoteMintDecimals
-    );
+    const maxRebalancingAmountNative = uiToNative(maxRebalancingAmount, depository.quoteMintDecimals);
+    const perpSide = polarity == PnLPolarity.Positive ? 'short' : 'long'; //'sell' : 'buy';
```
`===`

0xCryptoSheik commented on src/client.ts:
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
-    const collateralAmountBN = new BN(
-      collateralAmount * 10 ** depository.collateralMintDecimals
-    );
+    const userCollateralATA = await utils.token.associatedAddress({
```
wrap in a `Promise.all`

0xCryptoSheik commented on src/client.ts:
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
+    });
+    const collateralAmountBN = uiToNative(collateralAmount, depository.collateralMintDecimals);
+    const limit_price = (
+      await depository.getLimitPrice(
```
Ideally wrap in the `Promise.all` of above, don't bother otherwise

0xCryptoSheik commented on src/client.ts:
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
+        mango
+      )
+    ).toNumber();
+    console.log('LIMIT_PRICE for PERP ORDER : ', limit_price);
```
No logs for merge

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
-    )[0];
-    const redeemAmount = new BN(
-      amountRedeemable * 10 ** controller.redeemableMintDecimals
+    const userCollateralATA = await utils.token.associatedAddress({
```
Promise.all

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    });
+    const redeemAmountBN = uiToNative(amountRedeemable, controller.redeemableMintDecimals)
+    const limit_price = (
+      await depository.getLimitPrice(I80F48.fromNumber(slippage), 'long', mango)
```
Promise.all if possible

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
-    return this.instruction.redeemFromMangoDepository(redeemAmount, slippage, {
+  }
+
+  public async createRegisterZoDepositoryInstruction(
```
remove `async`

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
template string

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+      depository.collateralMintSymbol + '-PERP'
+    );
+    const openOrdersPda =
+      await depository.getZoAccountCollateralOpenOrdersAccountKey(zo, control);
```
in a `Promise.all` with the above

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const userCollateralATA = await utils.token.associatedAddress({
```
`Promise.all` all the promises that can be done concurrently

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    });
+    const limitPrice = depository.getLimitPrice(slippage, 'short', zo);
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
template string

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+      )
+    );
+    console.log('QUOTE LOT SIZE', market.decoded['quoteLotSize'].toNumber());
+    console.log('BASE LOT SIZE', market.decoded['baseLotSize'].toNumber());
```
no logs for merge

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+          )[0],
+          zoControl: marginAccount.control.pubkey,
+          zoOpenOrders:
+            await depository.getZoAccountCollateralOpenOrdersAccountKey(
```
avoid putting `await` of async operations in long-ass object declarations, prefer `await` in a variable and then use that later on

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const userCollateralATA = await utils.token.associatedAddress({
```
`Promise.all`

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    const perpPrice = depository.getPerpPriceUI(zo);
+    const limitPrice = depository.getLimitPrice(slippage, 'long', zo);
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
template string

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    const maxBaseQtyBn = new BN(
+      Math.round(maxQuoteQtyLotPurchaseBn.div(perpPriceBn).toNumber())
+    );
+    console.log('Limit price', limitPriceBn.toNumber());
```
no log for merge

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+          )[0],
+          zoControl: marginAccount.control.pubkey,
+          zoOpenOrders:
+            await depository.getZoAccountCollateralOpenOrdersAccountKey(
```
outside of the object declaration please

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+    options: ConfirmOptions
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const authorityQuoteATA = await utils.token.associatedAddress({
```
Promise.all

0xCryptoSheik commented on src/mango/depository.ts:
```diff
@@ -263,6 +280,62 @@ export class MangoDepository {
   }
 
+  // Return the depository lifetime funding PnL
+  public async getFundingPnl(): Promise<number> {
```
ok lol nice

0xCryptoSheik commented on src/mango/depository.ts:
```diff
@@ -263,6 +280,62 @@ export class MangoDepository {
+    let collateralDepositInterest: number = 0;
+    let mangoCache = await mango.getCache();
+
+    var response = await fetch(
```
Promise.all with the previous

0xCryptoSheik commented:
> BREAKING: Some transaction creation instructions now return a promise instead - I think this is bad because it's now blocking? What is the best way, inject the needed accounts?

It's fine, though it'd be good to check a little deeper as to why this changed & other potential impact

0xCryptoSheik commented:
> BREAKING: Upgrade anchor to 0.24.2 (from 0.22.0)

Ideally when we start using this, have everything aligned to this version of anchor (staking, FE, client, mango client, zo client, program).

0xCryptoSheik commented:
> @0xCryptoSheik How to handle these issues with zero_one dep, can we ignore? <img alt="Screenshot 2022-05-06 at 10 58 53" width="778" src="https://user-images.githubusercontent.com/40299747/167059637-e9fdd2ff-6e70-42e3-8b16-211e228afa34.png">

For now yes ignore, but ideally need to find the type definitions for bn.js, I'll look into it soon

0xCryptoSheik reviewed (changes requested):
See comments

0xCryptoSheik commented:
I'm surprised we don't already have the issue with BN, anchor must have provided their own declarations.

0xCryptoSheik commented:
Aight typechecking errors are fixed/suppressed, will make another pass before merge, don't hesitate to run `npm run validate` from time to time

CryptoMaxPlanck committed:
USDC_MINTING
- add instruction for mint and redeem in the client
- update IDL
- add helper getOffsetUnrealizedPnl
- npm run format

0xCryptoSheik commented:
> Add methods to retrieve Pnl for borrow/deposit/funding from Mango Stats in MangoDepository

@acamill can we add this as part of its own PR?

Sounds unrelated to the program update & 01 support & has a few considerations of its own.

0xCryptoSheik commented on src/01/depository.ts:
```diff
@@ -0,0 +1,183 @@
+    this.quoteMintSymbol = quoteMintName;
+    this.quoteMintDecimals = quoteMintDecimals;
+    this.pda = findAddrSync(
+      [Buffer.from('ZODEPOSITORY'), mint.toBuffer()],
```
`Buffer` is a Node.js API, it's not available in the browser, which means consumers who'll want to use `uxd-client` for an app, will need to tweak their bundler configuration to include a browser-compatible `Buffer` module & make it available for global use.

I didn't see we did that as well in other places, it's fine for now but better not to do this & include a buffer module dependency instead at some point.

0xCryptoSheik replied:
Leaving as-is for now, may fix in another PR

Commit:
USDC_MINTING
- Updating IDL
- Reflect updates in the MangoDepository interface

Commit:
misc
- fix npm run build

Commit:
- add mangoAccount to the quoteMint
- add mutable to some accounts in usdc mint instructions
- update the root and node bank to quote

CryptoMaxPlanck committed:
misc
- new idl
- why "M"ango? but added that again...
- quoteMint's mango accounts fixed

CryptoMaxPlanck committed:
- fix units of netQuoteMinted
- remove unnecessary commentrs

CryptoMaxPlanck committed:
- nothing, idk why this is a separate one

Commit:
Format + names

Commit:
Typo

Commit:
Fix maj in IDL account name

Commit:
Add tx-expect-error

Commit:
Fix error

acamill committed:
Merge pull request #7 from UXDProtocol/feature/usdc-minting

Feature/usdc minting

cnek committed:
updated idl

cnek committed:
added ix interface

0xCryptoSheik commented:
> > Add methods to retrieve Pnl for borrow/deposit/funding from Mango Stats in MangoDepository
> 
> @acamill can we add this as part of its own PR?
> 
> Sounds unrelated to the program update & 01 support & has a few considerations of its own.

Confirmed this, extracting & opening a new MR

cnek committed:
updated naming

cnek committed:
removed async

cnek committed:
Merge pull request #8 from UXDProtocol/disable_depository_mint

Disable depository mint support

Commit:
@uxd-protocol/uxd-client@2.0.0-beta.3

0xCryptoSheik committed:
remove Mango Depository & 01 updates (backport their own branches)

0xCryptoSheik commented:
removing everything related to 01 too

0xCryptoSheik commented on src/client.ts:
```diff
@@ -17,6 +18,7 @@ import { findATAAddrSync } from './utils';
 import NamespaceFactory from './namespace';
 import { IDL as UXD_IDL } from './idl';
 import { PnLPolarity } from './interfaces';
+import { I80F48 } from './index';
```
rather from Mango

0xCryptoSheik commented on src/client.ts:
```diff
@@ -128,12 +132,22 @@ export class UXDClient {
+
+    const [userCollateralATA, userQuoteATA, limitPriceI80F48] =
+      await Promise.all([
+        utils.token.associatedAddress({
```
@acamill so basically `findATAAddrSync` is deprecated in favor of ` utils.token.associatedAddress` ?

acamill replied:
I think it's the new interface in spl token 0.2.0 but I was accessing it through 01.. cheater.

I think we can roll back to the 0.1.8 way, I did in the program.

If I'm not mistaken.

0xCryptoSheik replied:
Mhh, you're accessing it from `@project-serum/anchor`.

I'll take a closer look at the differences between the two & migrate to the best.

0xCryptoSheik replied:
PSA: please do not use `utils.token.associatedAddress` from anchor, or any async `findProgramAddress` .
It was originally async in `@solana/web3.js`, then they changed it to be synchronous, kept the async version for backwards compatibility.
TLDR: use `PublicKey.findProgramAddressSync`
I will change it in this branch to use only the sync version, and remove any use of `utils.token.associatedAddress` from anchor

0xCryptoSheik commented on src/client.ts:
```diff
@@ -452,4 +526,102 @@ export class UXDClient {
+      user,
+      controller.redeemableMintPda
+    )[0];
+    const redeemableAmountNativeBN = new BN(
```
uiToNative ?

acamill replied:
Yes preferable anywhere this is used

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
+    const mangoPerpMarketConfig = mango.getPerpMarketConfig(
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
```
`findATAAddrSync` or `utils.token.associatedAddress` ?

There can be only one.

acamill replied:
FindAtaAddrSync let's not change what doesn't need to be change

0xCryptoSheik replied:
https://github.com/UXDProtocol/uxd-client/pull/3#discussion_r875697863 if that's ok

acamill replied:
Yes! I did change some stuff that wasn't needed, I mean that was the meaning of my "let's not change..." above 😁 
Thanks for the keen eye

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
+    return this.instruction.quoteMintWithMangoDepository(quoteAmountNativeBN, {
       accounts: {
-        user,
+        user: user,
```
just `user`, is good

0xCryptoSheik commented on src/client.ts:
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
     )[0];
-    const redeemAmount = new BN(
-      amountRedeemable * 10 ** controller.redeemableMintDecimals
+    const quoteAmountNativeBN = new BN(
```
`uiToNative` ?

0xCryptoSheik commented on src/client.ts:
```diff
@@ -452,4 +526,102 @@ export class UXDClient {
+    const mangoPerpMarketConfig = mango.getPerpMarketConfig(
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
```
findATAAddrSync vs utils.token.blah

0xCryptoSheik commented on src/interfaces.ts:
```diff
@@ -34,7 +34,24 @@ export class MangoDepositoryAccountV1 {
 }
 
-// V2
+// V2 with ZO no USDC
```
Can we haz better comments please?

0xCryptoSheik commented on src/mango/depository.ts:
```diff
@@ -290,4 +292,35 @@ export class MangoDepository {
+  }
+
+  // The side of the taker (the user)
+  async getLimitPrice(
```
assuming these 2 methods are now required for 3.1.0

acamill replied:
Yes!

0xCryptoSheik committed:
misc: dependencies, unnecessarily async, findProgramAddress

Commit:
- remove zo version of mangoDepositoryAccount
- import I80F48 from mango instead of self reference

acamill commented:
@0xCryptoSheik Thanks for handling this, pushed the last tiny changes

acamill committed:
Add support for quote_mint_redeem_softcap

acamill committed:
Fix greg feedback

acamill committed:
Fix

acamill committed:
Merge branch 'prog/v3.1.0' into ix/quote-soft-cap

acamill committed:
Shorthand

0xCryptoSheik commented:
poke @acamill 

```
src/mango/depository.ts
[error] src/mango/depository.ts: SyntaxError: Declaration or statement expected. (344:3)
[error]   342 |   // `collateralQuantity` : In Collateral UI units
[error]   343 |   //  Output -> breakdown of estimated costs
[error] > 344 |   public getMintingEstimates(
[error]       |   ^
[error]   345 |     collateralQuantity: number,
[error]   346 |     perpPrice: number,
[error]   347 |     mintingPriceImpact: number,
```

0xCryptoSheik committed:
fixup

0xCryptoSheik commented:
poke @acamill 

```
 FAIL  src/__tests__/mango-price-impact.spec.ts
  Price Impact / Mint/Redeem estimations tests
    ✕ Minting (138 ms)
    ✕ Redeeming (73 ms)

  ● Price Impact / Mint/Redeem estimations tests › Minting

    Price impact couldn't be determined.

      424 |     const priceImpact = sideBook.getImpactPriceUi(nativeQuantity);
      425 |     if (typeof priceImpact === 'undefined') {
    > 426 |       throw new Error("Price impact couldn't be determined.");
          |             ^
      427 |     }
      428 |     return priceImpact;
      429 |   }

      at MangoDepository.getMangoPerpOrderBookPriceImpact (src/mango/depository.ts:426:13)
      at Object.<anonymous> (src/__tests__/mango-price-impact.spec.ts:29:32)
```

cnek committed:
updated naming and idl

cnek committed:
fixed format

cnek committed:
updated idl

cnek committed:
updated ix creation and on chain ac interface

acamill committed:
Merge pull request #13 from UXDProtocol/fix/info-6

updated "disable minting" naming and idl

cnek committed:
Merge branch 'prog/v3.1.0' of https://github.com/UXDProtocol/uxd-client into ix/quote-soft-cap

cnek committed:
Merge pull request #11 from UXDProtocol/ix/quote-soft-cap

Add support for quote_mint_redeem_softcap

CryptoMaxPlanck committed:
Add getOffsetUnrealizedPnl to depository (#12)

* re-adding

* fixed unit conversion and merged in softCap

* fixed idl format

* removed comment

* Update interfaces.ts

Co-authored-by: cnek <ctk2012ac@gmail.com>
Co-authored-by: acammm <40299747+acamill@users.noreply.github.com>

cnek committed:
rm price impact test (#14)

0xCryptoSheik committed:
chore: optimize concurrent async operations

0xCryptoSheik committed:
chore: pass with no tests

0xCryptoSheik committed:
chore: use findMultipleATAAddSync when possible

cnek committed:
revert "redeem only mode"

0xCryptoSheik committed:
fixup

0xCryptoSheik committed:
fixup

0xCryptoSheik committed:
fixup
